### ISSUE TITLE
[ENH] Replace sklearn GradientBoostingRegressor by lightgbm

### ISSUE THREAD

**Issue by ThomasBury** [scikit-optimize/scikit-optimize#922]
Current Situation
=================
The gbrt optimizer uses the sklearn gradient boosting

Enhancement
===========
Replace the sklearn gradient boosting by a lightgbm

Reasoning
=========
This would enhance the running time to a running time comparable to hyperopt (I had similar perf and running than hyperopt on a test classification problem, more extensive testing might be necessary). Moreover, it seems to be more accurate.

Implementation
==============
By setting:

```python
optimizer_kwargs={'base_estimator': LGBMQuantileRegressor(), 'acq_optimizer': "sampling"}
```

with 

```python
import numpy as np

from sklearn.base import clone
from sklearn.base import BaseEstimator, RegressorMixin
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.utils import check_random_state
from joblib import Parallel, delayed
from lightgbm import LGBMRegressor


def _parallel_fit(regressor, X, y):
    return regressor.fit(X, y)


class LGBMQuantileRegressor(BaseEstimator, RegressorMixin):
    """Predict several quantiles with one estimator.
    This is a wrapper around `LGBMRegressor`'s quantile
    regression that allows you to predict several `quantiles` in
    one go.
    Parameters
    ----------
    quantiles : array-like
        Quantiles to predict. By default the 16, 50 and 84%
        quantiles are predicted.
    base_estimator : LGBMRegressor instance or None (default)
        Quantile regressor used to make predictions. Only instances
        of `GradientBoostingRegressor` are supported. Use this to change
        the hyper-parameters of the estimator.
    n_jobs : int, default=1
        The number of jobs to run in parallel for `fit`.
        If -1, then the number of jobs is set to the number of cores.
    random_state : int, RandomState instance, or None (default)
        Set random state to something other than None for reproducible
        results.
    """

    def __init__(self, quantiles=[0.16, 0.5, 0.84],
                 n_jobs=1, random_state=None):
        self.quantiles = quantiles
        self.random_state = random_state
        self.n_jobs = n_jobs

    def fit(self, X, y):
        """Fit one regressor for each quantile.
        Parameters
        ----------
        X : array-like, shape=(n_samples, n_features)
            Training vectors, where `n_samples` is the number of samples
            and `n_features` is the number of features.
        y : array-like, shape=(n_samples,)
            Target values (real numbers in regression)
        """
        rng = check_random_state(self.random_state)
        base_estimator = LGBMRegressor(objective='quantile')
        # The predictions for different quantiles should be sorted.
        # Therefore each of the regressors need the same seed.
        base_estimator.set_params(random_state=42)
        regressors = []
        for q in self.quantiles:
            regressor = clone(base_estimator)
            regressor.set_params(alpha=q)

            regressors.append(regressor)

        self.regressors_ = Parallel(n_jobs=self.n_jobs, backend='threading')(
            delayed(_parallel_fit)(regressor, X, y)
            for regressor in regressors)

        return self

    def predict(self, X, return_std=False, return_quantiles=False):
        """Predict.
        Predict `X` at every quantile if `return_std` is set to False.
        If `return_std` is set to True, then return the mean
        and the predicted standard deviation, which is approximated as
        the (0.84th quantile - 0.16th quantile) divided by 2.0
        Parameters
        ----------
        X : array-like, shape=(n_samples, n_features)
            where `n_samples` is the number of samples
            and `n_features` is the number of features.
        """
        predicted_quantiles = np.asarray(
            [rgr.predict(X) for rgr in self.regressors_])
        if return_quantiles:
            return predicted_quantiles.T

        elif return_std:
            std_quantiles = [0.16, 0.5, 0.84]
            is_present_mask = np.in1d(std_quantiles, self.quantiles)
            if not np.all(is_present_mask):
                raise ValueError(
                    "return_std works only if the quantiles during "
                    "instantiation include 0.16, 0.5 and 0.84")
            low = self.regressors_[self.quantiles.index(0.16)].predict(X)
            high = self.regressors_[self.quantiles.index(0.84)].predict(X)
            mean = self.regressors_[self.quantiles.index(0.5)].predict(X)
            return mean, ((high - low) / 2.0)

        # return the mean
        return self.regressors_[self.quantiles.index(0.5)].predict(X)
```

## Test

```python
from functools import partial
import time
from skopt import gp_minimize, forest_minimize, dummy_minimize, gbrt_minimize
from skopt.plots import plot_convergence

func = partial(branin, noise_level=2.0)
bounds = [(-5.0, 10.0), (0.0, 15.0)]
n_calls = 60

def run(minimizer, n_iter=5):
    return [minimizer(func, bounds, n_calls=n_calls, random_state=n)
            for n in range(n_iter)]

# Random search
tic = time.time()
dummy_res = run(dummy_minimize)
print("RND running time: {0:2.2f} s".format(time.time()-tic))

# Gaussian processes
tic = time.time()
gp_res = run(gp_minimize)
print("GP running time: {0:2.2f} s".format(time.time()-tic))

# Random forest
tic = time.time()
rf_res = run(partial(forest_minimize, base_estimator="RF"))
print("RF running time: {0:2.2f} s".format(time.time()-tic))

# Gradient boosting
tic = time.time()
gbrt_res = run(gbrt_minimize)
print("GBRT running time: {0:2.2f} s".format(time.time()-tic))

# Lightgbm
tic = time.time()
lgb_res = run(partial(gbrt_minimize, base_estimator=LGBMQuantileRegressor(), acq_optimizer= "sampling"))
print("LGB running time: {0:2.2f} s".format(time.time()-tic))

fig = plt.figure(figsize=(8, 6))
plot = plot_convergence(("dummy_minimize", dummy_res),
                        ("gp_minimize", gp_res),
                        ("forest_minimize('rf')", rf_res),
                        ("gbrt_minimize", gbrt_res),
                        ("lgb_minimize", lgb_res),
                        true_minimum=0.397887, yscale="log")

plot.legend(loc="best", prop={'size':10}, numpoints=1)
```

with outputs:

![image](https://user-images.githubusercontent.com/28918551/86931491-772d2c80-c138-11ea-85fa-718718e88a39.png)
![image](https://user-images.githubusercontent.com/28918551/86931540-844a1b80-c138-11ea-9bac-aa59389394d0.png)

Tasks
=====
- Either replace the GradientBoostingQuantileRegressor
- Or keeping both like for ET and RF, making LGBM the default
